File: algorithms/delphi/sorting/QuickSort.pas
program QuickSort;
{$APPTYPE CONSOLE}
uses SysUtils;

type
  TIntegerArray = array of Integer;

procedure QuickSortRec(var Arr: TIntegerArray; Low, High: Integer);
var
  i, j, pivot, temp: Integer;
begin
  i := Low;
  j := High;
  pivot := Arr[(Low + High) div 2];
  
  repeat
    while Arr[i] < pivot do Inc(i);
    while Arr[j] > pivot do Dec(j);
    
    if i <= j then
    begin
      temp := Arr[i];
      Arr[i] := Arr[j];
      Arr[j] := temp;
      Inc(i);
      Dec(j);
    end;
  until i > j;
  
  if Low < j then QuickSortRec(Arr, Low, j);
  if i < High then QuickSortRec(Arr, i, High);
end;

var
  Arr: TIntegerArray;
  i: Integer;
begin
  SetLength(Arr, 6);
  Arr[0] := 10; Arr[1] := 7; Arr[2] := 8; 
  Arr[3] := 9;  Arr[4] := 1; Arr[5] := 5;
  
  Write('Original: ');
  for i := 0 to High(Arr) do Write(IntToStr(Arr[i]) + ' ');
  Writeln;
  
  QuickSortRec(Arr, 0, High(Arr));
  
  Write('Sorted: ');
  for i := 0 to High(Arr) do Write(IntToStr(Arr[i]) + ' ');
  Writeln;
end.

File: algorithms/mumps/sorting/quick_sort.m
QSORT
    ; MUMPS QuickSort
    NEW ARR
    SET ARR(1)=10, ARR(2)=7, ARR(3)=8, ARR(4)=9, ARR(5)=1, ARR(6)=5
    
    WRITE "Original: "
    FOR I=1:1:6 WRITE ARR(I)," "
    WRITE !
    
    DO QUICK(.ARR, 1, 6)
    
    WRITE "Sorted:   "
    FOR I=1:1:6 WRITE ARR(I)," "
    WRITE !
    QUIT

QUICK(A, LEFT, RIGHT)
    NEW I,J,PIVOT,TEMP
    IF LEFT'<RIGHT QUIT
    SET PIVOT=A(RIGHT)
    SET I=LEFT-1
    FOR J=LEFT:1:RIGHT-1 DO
    . IF A(J)<PIVOT DO
    . . SET I=I+1
    . . SET TEMP=A(I), A(I)=A(J), A(J)=TEMP
    SET TEMP=A(I+1), A(I+1)=A(RIGHT), A(RIGHT)=TEMP
    
    DO QUICK(.A, LEFT, I)
    DO QUICK(.A, I+2, RIGHT)
    QUIT
